cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java

```diff
@@ -2882,170 +2869,159 @@ StatefulSet getKafkaStatefulSet()   {
-         * to not happen in parallel or have a risk that there will be several pods deleted at the same time (which can
-         * affect availability). If multiple pods are marked for deletion, the next one will be deleted in the next loop.
+         * @param ctlrResourceName  Name of the controller resource (e.g. StatefulSet)
+         * @param selector          Selector for selecting the Pods belonging to this controller
+         * @param desiredPvcs       The list of desired PVCs which should be created after the old Pod and PVCs are deleted
          *
-         * @param stsOperator           StatefulSet Operator for managing stateful sets
-         * @param sts                   StatefulSet which owns the pods which should be checked for deletion
-         * @param desiredPvcs           The list of PVCs which should exist
-         * @param existingPvcsFuture    Future which will return a list of PVCs which actually exist
-         * @return
+         * @return                  Future indicating the result of the cleanup. Returns always success if there are no
+         * pods to cleanup.
          */
-        Future<Void> maybeCleanPodAndPvc(StatefulSetOperator stsOperator, StatefulSet sts, List<PersistentVolumeClaim> desiredPvcs, Future<List<PersistentVolumeClaim>> existingPvcsFuture)  {
-            if (sts != null) {
-                LOGGER.debugCr(reconciliation, "Considering manual cleaning of Pods for StatefulSet {}", sts.getMetadata().getName());
-
-                String stsName = sts.getMetadata().getName();
-
-                for (int i = 0; i < sts.getSpec().getReplicas(); i++) {
-                    String podName = stsName + "-" + i;
-                    Pod pod = podOperations.get(namespace, podName);
-
-                    if (pod != null) {
-                        if (Annotations.booleanAnnotation(pod, AbstractScalableResourceOperator.ANNO_STRIMZI_IO_DELETE_POD_AND_PVC, false)) {
-                            LOGGER.debugCr(reconciliation, "Pod and PVCs for {} should be deleted based on annotation", podName);
-
-                            return existingPvcsFuture
-                                    .compose(existingPvcs -> {
-                                        List<PersistentVolumeClaim> deletePvcs;
+        Future<ReconciliationState> maybeManualPodCleaning(String ctlrResourceName, Labels selector, List<PersistentVolumeClaim> desiredPvcs) {
+            return podOperations
+                    .listAsync(namespace, selector)
+                    .compose(pods -> {
+                        // Only one pod per reconciliation is rolled
+                        Pod podToClean = pods
+                                .stream()
+                                .filter(pod -> Annotations.booleanAnnotation(pod, AbstractScalableResourceOperator.ANNO_STRIMZI_IO_DELETE_POD_AND_PVC, false))
+                                .findFirst()
```

Comment:
Just a question. Why can't we get all such pods and execute all the operations in single reconciliation?

Reply:
I really followed the previous logic here. IIRC, the reason why it was implemented this way was to avoid causing too much disruption by deleting too many things at the same time.

Reply:
Ah, thank you.